### PR TITLE
[C++ worker] Ray normal task for RAY_REMOTE

### PR DESCRIPTION
**Pull request by qicosmos**
## Why are these changes needed?
We should support safe c++ worker API according the [doc](https://docs.google.com/document/d/1pBRUm-w2LKXTte8uXmvQ23l7gOJV2IOwirCTPjtPGWk/edit#heading=h.86fui9aqjkh6), this pr supports normal task for RAY_EMOTE.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by SongGuyang** [comment]
@qicosmos Can you add a detail description? We already support normal task, you title will confuse others.

File: cpp/include/ray/api/exec_funcs.h [hunk 1]
```diff
@@ -15,9 +15,50 @@
...
+      result = (*func_ptr)(args_buffer);
+    } catch (const std::exception &ex) {
+      result = PackError(ex.what());
+    }
+  } while (0);
+
+  return result;
+}
+
+inline static msgpack::sbuffer CallInDll(
```
Comment (SongGuyang) [inline cpp/include/ray/api/exec_funcs.h:55]: Why need this?

File: cpp/include/ray/api/object_ref.h [hunk 2]
```diff
@@ -93,5 +104,59 @@ template <typename T>
 inline std::shared_ptr<T> ObjectRef<T>::Get() const {
   return GetFromRuntime(*this);
 }
+
+template <>
+class ObjectRef<void> {
```
Comment (SongGuyang) [inline cpp/include/ray/api/object_ref.h:109]: Can we use only one class define? This bring more redundant code.

File: cpp/src/ray/runtime/task/native_task_submitter.cc [hunk 3]
```diff
@@ -9,6 +9,12 @@ namespace ray {
 namespace api {
 
 RayFunction BuildRayFunction(InvocationSpec &invocation) {
+  if (ray::api::RayConfig::GetInstance()->use_ray_remote) {
+    auto function_descriptor =
+        FunctionDescriptorBuilder::BuildCpp(invocation.lib_name, "", "");
```
Comment (SongGuyang) [inline cpp/src/ray/runtime/task/native_task_submitter.cc:14]: We need put function name to the function descriptor.

File: src/ray/util/logging.cc [hunk 4]
```diff
@@ -104,22 +104,28 @@ inline const char *ConstBasename(const char *filepath) {
 
 /// A logger that prints logs to stderr.
 /// This is the default logger if logging is not initialized.
+/// NOTE(lingxuan.zlx): Default stderr logger must be singleton and global
+/// variable so core worker process can invoke `RAY_LOG` in its whole lifecyle.
 class DefaultStdErrLogger final {
  public:
+  std::shared_ptr<spdlog::logger> GetDefaultLogger() { return default_stderr_logger_; }
```
Comment (SongGuyang) [inline src/ray/util/logging.cc:111]: Why need this change in this PR?

**Comment by AmeerHajAli** [comment]
@qicosmos , what is the status of this?

**Comment by SongGuyang** [comment]
I found we have lost the tests of old version(use_ray_remote = false). Maybe we should test both `use_ray_remote = true` and `use_ray_remote = false` based on shared test cases.

**Comment by SongGuyang** [comment]
@AmeerHajAli  some issues are being fixed. We will merge this PR this week.

**Review by SongGuyang — APPROVED** [review]
LGTM

**Pull request merged**